README.md

```diff
@@ -1,25 +1,67 @@
-Lager graylog udp
-=================
+## lager_graylog
 
-There are two modules:
-* lager\_gelf\_formatter, which will format your logs according to gelf
-standards
-* graylog\_udp\_sender, which will send your logs to graylog instance
+This application provides lager backends for sending log messages to [Graylog](https://www.graylog.org/)
+over UDP or TCP, and the formatter module which spits out logs in [GELF](http://docs.graylog.org/en/stable/pages/gelf.html)
+format.
 
-Config
-======
+### TCP backend
 
+When you need a reliable log delivery, you can use the backend which connects to Graylog using TCP.
+This backend will also try to reconnect to Graylog indefinitely (with backoff) in case of connection
+failure.
+
+To use it, just declare it in your lager config:
+
+```erlang
+[{lager, [{handlers, [{lager_graylog_tcp_backend, [{host, "graylog-hostname"},
+                                                   {port, 12201}]
+
+}]}]}].
 ```
-[{lager, [
-    {handlers, [
-        {lager_udp_backend,
-            [info,
-            {host, "graylog.erlang-solutions.com"},
-            {port, 12201},
-            {level, info},
-            {formatter, lager_gelf_formatter},
-            {formatter_config, [{metadata, [{service, "SERVICE NAME"}]}]}
-            ]}
-        ]}
-    ]}].
+
+This backend accepts configuration options as a list of tuples with atom keys. The following
+options are supported:
+
+* `host` (**required**) - host name or IP address (basically everything accepted by `gen_tcp:connect/3`)
+   of the Graylog instance
+* `port` (**required**) - port where the Graylog instance accepts TCP connection
+* `level` (**optional**, default: `info`) - the log level (any log level supported by lager)
+* `formatter` (**optional**, default: `lager_graylog_gelf_formatter`) - the formatter used for
+  formatting log messages before sending to Graylog
+* `formatter_config` - (**optional**, default: `[]`) - passed as a second argument to formatter's
+  `format/2` function
+
+### GELF formatter
+
+The GELF formatter is implemented by the `lager_graylog_gelf_formatter` module. It formats log
+messages according to GELF version 1.1. The following fields are always included in the message:
+
+* `"version"` - always has value `"1.1"`
+* `"host"` - hostname retrieved using `inet:gethostname/0` (might be overriden)
+* `"short_message"` - the log message
+* `"level"` - the log severity formatted as a number as in [syslog](https://en.wikipedia.org/wiki/Syslog#Severity_level)
+
+In addition, all metadata provided by lager in the log message will be included as additional fields
+(thus prefixed with `_`). Please note that this formatter can only format binaries, strings, atoms,
+integers, floats (always formatted with 6 decimal places), references, ports and pids.
```

Comment:
What will happen to maps?

Reply:
This formatter can't format maps. The easiest solution to that is to use a JSON library to encode the payload, since GELF is just a JSON. We do exactly that in the Elixir Logger Graylog library.

However, I never know which JSON encoder to choose in the Erlang world - `jsx` slower than `jiffy`, but it doesn't include native code which IMO is a plus. Maybe we should allow users to configure the encoder?

Reply:
I'd be for jiffy because of performance. Logs may be formatted very often so performance matters here. Does GELF and Graylog allow such nested structure at all? Maybe maps have to flatten in a similar way to the builtin lager console formatter?

Reply:
I'd be for pluggable encoder, because someone might build for some strange target and won't be able to compile jiffy. In addition, if they're already using different JSON library, they won't have another one as a dependency.


Anyway, that's a subject for another PR.

Reply:
Right now the formatter will crash on maps so let's not worry about what Graylog allows to do.

Reply:
Oh, and I agree on the performance aspect of jiffy, of course!